### PR TITLE
[Baseline][Link LightGBM] Miscellaneous improvement for LightGBM model

### PR DESCRIPTION
**Pull request by XinweiHe**
### Summary
1. Using past visit and global popularity heuristics as candidate generation for validation data as well as test set, so that we are able to obtain the link prediction metrics for validation set.
2.  Add a helper function to include global popularity and past visit as training features

### Testing
Updated benchmarking results

Screenshot 2024-06-13 at 11 40 08 AM

### CODE REVIEW

File: examples/lightgbm_link_baseline.py [hunk 1]
```diff
@@ -54,12 +54,12 @@
...
-]
-dst_entity_table_col_to_stype = dataset2inferred_stypes[args.dataset][
-    task.dst_entity_table
-]
+src_entity_table_col_to_stype = copy.deepcopy(
+    dataset2inferred_stypes[args.dataset][task.src_entity_table]
+)
+dst_entity_table_col_to_stype = copy.deepcopy(
+    dataset2inferred_stypes[args.dataset][task.dst_entity_table]
+)
```
Comment (weihua916) [inline examples/lightgbm_link_baseline.py:62]: Could you clarify why deep copy is required here? `dataset2inferred_stypes[args.dataset]` is never used later, so it seems fine to modify it in-place.

File: examples/lightgbm_link_baseline.py [hunk 2]
```diff
@@ -86,9 +86,13 @@
 
 # randomly subsample in case training data size is too large.
 sampled_train_table = copy.deepcopy(train_table)
+sampled_val_table = copy.deepcopy(val_table)
 if args.sample_size > 0 and args.sample_size < len(sampled_train_table):
     sampled_idx = np.random.permutation(len(sampled_train_table))[: args.sample_size]
     sampled_train_table.df = sampled_train_table.df.iloc[sampled_idx]
+if args.sample_size > 0 and args.sample_size < len(sampled_val_table):
+    sampled_idx = np.random.permutation(len(sampled_val_table))[: args.sample_size]
+    sampled_val_table.df = sampled_val_table.df.iloc[sampled_idx]
```
Comment (weihua916) [inline examples/lightgbm_link_baseline.py:95]: I don't want validation data to be sampled because we report validation performance in our paper.
Also, given that we use full test data anyway, the speedup is not that large.

Comment (XinweiHe) [reply examples/lightgbm_link_baseline.py:95]: I see, reverting this now.

File: examples/lightgbm_link_baseline.py (hunk 1 above)
Comment (XinweiHe) [reply examples/lightgbm_link_baseline.py:62]: Because there are cases where `task.src_entity_table == task.dst_entity_table` which makes `src_entity_table_col_to_stype` and `dst_entity_table_col_to_stype` points to the same object, which will result in error in downstream logics.

Comment (rusty1s) [reply examples/lightgbm_link_baseline.py:62]: What is the error? Are we modifying this dict in-place somewhere? That's probably something we wanna do.

File: examples/lightgbm_link_baseline.py [hunk 3]
```diff
@@ -90,6 +90,75 @@
...
+    )
+
+    # Fill NaN values with 0 (if there are any right_entity in train_table_df not present in past_table_df)
+    train_table_df["num_past_visit"] = (
+        train_table_df["num_past_visit"].fillna(0).astype(int)
+    )
+
+    # Add percentage of global popularity for each right_entity
+    # Count occurrences of each right_entity
+    right_entity_count = exploded_past_table[right_entity].value_counts().reset_index()
```
Comment (rusty1s) [inline examples/lightgbm_link_baseline.py:139]: `dst` instead of `right`.

File: examples/lightgbm_link_baseline.py [hunk 4]
```diff
@@ -90,6 +90,75 @@
...
+
+    # Calculate the fraction
+    total_right_entities = len(exploded_past_table)
+    right_entity_count["global_popularity_fraction"] = (
+        right_entity_count["count"] / total_right_entities
```
Comment (rusty1s) [inline examples/lightgbm_link_baseline.py:144]: divide by `count.max()`?

**Review by rusty1s — APPROVED** [review]

File: examples/lightgbm_link_baseline.py (hunk 1 above)
Comment (XinweiHe) [reply examples/lightgbm_link_baseline.py:62]: When src table and dst table are the same table the error will arise. e.g. in rel-stack post-post-related task
